### PR TITLE
Forbid multiple calls to initialize

### PR DESCRIPTION
**Pull request by leodasvacas**
Remove `pub` from `global_registry`. `get_registry_with_config` renamed to `init_global_registry`. Forbid multiple calls to `intialize`, if the configurations were equal we allowed calls after the first as a no-op. However that's bogus because `Configuration` does not implement `PartialEq` and we would incorrectly assume that if the thread_num is equal then the configuration is equal.

### CODE REVIEW

File: rayon-core/src/configuration.rs [hunk 1]
```diff
@@ -152,26 +152,12 @@ impl Configuration {
...
-
-    let registry = registry::get_registry_with_config(config);
-
-    if let Some(value) = num_threads {
-        if value != registry.num_threads() {
-            return Err(InitError::GlobalPoolAlreadyInitialized);
-        }
-    }
-
+    config.validate()?;
```
Comment (nikomatsakis) [inline rayon-core/src/configuration.rs]: Unfortunately, we currently try and maintain compatibility with Rust 1.12.0, which I do not think had `?`; therefore it's better to use `try!`

**Comment by nikomatsakis** [comment]
👍 to this change overall; I agree we shouldn't assume that `Configuration` can be `PartialEq` (now that it has closures, that doesn't really make sense)

**Comment by leodasvacas** [comment]
Glad to know the change is welcome, replaced all question marks with try! including some in rayon-demo for consistency.

**Comment by nikomatsakis** [comment]
@leodasvacas it seems the tests fail:

```
thread 'join::test::sort' panicked at 'expected InitOk, but got GlobalPoolAlreadyInitialized', rayon-core/src/join/test.rs:48
```

you can probably just remove that initialization call, seems useless there

**Comment by leodasvacas** [comment]
@nikomatsakis Thanks for the heads up, I had also broken some benches, all green now.

**Comment by nikomatsakis** [comment]
Looking good, thank you.

**Pull request merged**